### PR TITLE
[Instrumentation] Do not run sanitizers for naked functions

### PR DESCRIPTION
**Pull request by antoniofrighetto**
Sanitizers instrumentation may be incompatible with naked functions, which lack of standard prologue/epilogue.

### CODE REVIEW

**Comment by fmayer** [comment]
What about HWAddressSanitizer?

**Comment by antoniofrighetto** [comment]
> What about HWAddressSanitizer?

Added!

**Comment by fmayer** [comment]
I am not sure the commit message is completely accurate though. IR docs say

> This attribute disables prologue / epilogue emission for the function. This can have very system-specific consequences. The arguments of a naked function can not be referenced through IR values.

This means _non arguments_ could still be referenced through IR.

It would probably help if you could point towards a specific problem that this is solving.

**Comment by antoniofrighetto** [comment]
See downstream issue: https://github.com/rust-lang/rust/issues/129224. Yeah, the commit message can be made more accurate. I was having in mind Rust Naked functions when I wrote it; indeed inline asm only is not a strict requirement in LLVM. That said, I'm not sure you'd be able to leverage Asan instrumentation properly (e.g., for stack accesses), if the function lacks of prologue/epilogue in the first place.

**Comment by fmayer** [comment]
LGTM % a more accurate commit message

**Comment by vitalybuka** [comment]
I suspect msan/dfsan still needs to set shadow for output value.
asan/hwasan probably has nothing to touch a naked function anyway.

**Comment by vitalybuka** [comment]
also msan instruments inline assembly, I think we still need it for naked function in general

File: llvm/test/Instrumentation/sanitizers-no-naked.ll [hunk 1]
```diff
@@ -0,0 +1,60 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes=asan -S | FileCheck %s
+; RUN: opt < %s -passes=tsan -S | FileCheck %s
+; RUN: opt < %s -passes=hwasan -S | FileCheck %s --check-prefixes=CHECK-HWASAN
```
Comment (vitalybuka) [inline llvm/test/Instrumentation/sanitizers-no-naked.ll:4]: there is no `-passes=msan`

**Comment by antoniofrighetto** [comment]
Tests in a separate commit, rephrased commit, dropped msan/dfsan. Does this look better?

File: llvm/test/Instrumentation/sanitizers-naked.ll [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+; RUN: opt < %s -passes=msan -S | FileCheck %s --check-prefixes=CHECK-MSAN
+; RUN: opt < %s -passes=hwasan -S | FileCheck %s --check-prefixes=CHECK-HWASAN
+; RUN: opt < %s -passes=dfsan -dfsan-track-origins=1  -S | FileCheck %s --check-prefixes=CHECK-DFSAN
+; RUN: opt < %s -passes='module(sancov-module)' -sanitizer-coverage-level=3 -sanitizer-coverage-control-flow -S | FileCheck %s --check-prefixes=CHECK-SANCOV
+; RUN: opt < %s -passes='module(sanmd-module)' -sanitizer-metadata-atomics -S | FileCheck %s --check-prefixes=CHECK-SANMD
+
+target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-unknown-linux-gnu"
+
+define void @naked_function() naked {
```
Comment (nikic) [inline llvm/test/Instrumentation/sanitizers-naked.ll:13]: Can you please add a byval argument to these functions? I think that will make it more likely that something actually gets instrumented and results in a verifier error.

Comment (antoniofrighetto) [reply llvm/test/Instrumentation/sanitizers-naked.ll:13]: May be missing something, how is the argument supposed to lead that something gets instrumented if the argument cannot be used?

Comment (nikic) [reply llvm/test/Instrumentation/sanitizers-naked.ll:13]: See https://llvm.godbolt.org/z/K4WdqTdve. Without the byval argument there is nothing to instrument anyway, so you're not really testing anything.

Comment (antoniofrighetto) [reply llvm/test/Instrumentation/sanitizers-naked.ll:13]: Doesn't your function miss naked attribute? The verifier would lead to an error in that case.

Comment (nikic) [reply llvm/test/Instrumentation/sanitizers-naked.ll:13]: Yes, that's the point. You want a test case that leads to a verifier error before your change and no longer causes one after your change.

Comment (antoniofrighetto) [reply llvm/test/Instrumentation/sanitizers-naked.ll]: OK I see it now, thanks. Should be better. I was slightly confused by the naked_function without being naked.

File: llvm/test/Instrumentation/sanitizers-naked.ll [hunk 3]
```diff
@@ -0,0 +1,89 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes=asan -S | FileCheck %s --check-prefixes=CHECK-ASAN
```
Comment (nikic) [inline llvm/test/Instrumentation/sanitizers-naked.ll:2]:
```suggestion
; RUN: opt < %s -passes=asan -S | FileCheck %s --check-prefixes=CHECK,CHECK-ASAN
```
etc. to remove duplicate check lines.

Comment (antoniofrighetto) [reply llvm/test/Instrumentation/sanitizers-naked.ll]: Added, but they seem unused.

Comment (nikic) [reply llvm/test/Instrumentation/sanitizers-naked.ll]: Hm okay, sounds like a bug in UTC.

**Comment by antoniofrighetto** [comment]
@nikic Think this is OK to land for you too?

**Review by nikic — APPROVED** [review]
LGTM

**Comment by antoniofrighetto** [comment]
Thanks!

**Pull request merged**